### PR TITLE
Fix validation for achievement conditions

### PR DESCRIPTION
**Pull request by minhtule**
- Fix the query to retrieve achievement conditions of a particular conditional object
- Improve the naming
- Remove complicated trait in achievement condition factory

### CODE REVIEW

**Comment by minhtule** [comment]
@allenwq does this reflect correctly the behavior of the function?

**Comment by allenwq** [comment]
@minhtu Yep, think it can be generalised.

And can we have a spec for it ? to test that it works for any conditionals.

**Comment by lowjoel** [comment]
I'm not entirely sure what's being fixed?

File: app/models/course/condition/achievement.rb [hunk 1]
```diff
@@ -20,11 +20,12 @@ def satisfied_by?(course_user)
 
   private
 
-  # Given an achievement, returns all its achievement conditions.
+  # Given a conditional object, returns all achievements that it requires.
   #
-  # @param [Course::Achievement] achievement The achievement to return achievement conditions for.
+  # @param [Object] conditional The object that is declared as acts_as_conditional and for which
```
Comment (lowjoel) [inline app/models/course/condition/achievement.rb]: You should specify the module that all such classes implement.

Comment (minhtule) [inline app/models/course/condition/achievement.rb]: `acts_as_conditional` is defined under 
```
module Extensions::Conditional::ActiveRecord::Base
  module ClassMethods
    # Functions from conditional-and-condition framework.
    # Declare this function in the conditional model that requires conditions.
    def acts_as_conditional
        ...
    end
  end
end
```

Any class subclassing `ActiveRecord::Base` can add `acts_as_conditional`. What module should I specify? `ActiveRecord::Base`? And how should I specify it in the documentation part?

**Comment by minhtule** [comment]
The main fix is in the SQL [query](https://github.com/Coursemology/coursemology2/pull/693/files#diff-bf3adb0382fd1dc7ad1781f2636a1007R41).

File: app/models/course/condition/achievement.rb (hunk 1 above)
Comment (lowjoel) [inline app/models/course/condition/achievement.rb]: Extensions::Conditiona...ConditionalInstanceMethods OR use duck typing. What are the methods each of these classes need to implement, and declare those methods. See the yard documentation.

File: app/models/course/condition/achievement.rb [hunk 2]
```diff
@@ -37,7 +38,9 @@ def achievement_conditions(achievement)
         (SELECT cca.achievement_id
           FROM course_condition_achievements cca INNER JOIN course_conditions cc
             ON cc.actable_type = 'Course::Condition::Achievement' AND cc.actable_id = cca.id
-            WHERE cc.conditional_id = #{achievement.id}) ids
+            WHERE cc.conditional_id = #{conditional.id}
```
Comment (lowjoel) [inline app/models/course/condition/achievement.rb]: Put this in a SQL heredoc block.

File: spec/models/course/condition/achievement_spec.rb [hunk 3]
```diff
@@ -20,12 +22,26 @@
 
       context "when an achievement is already included in its conditional's conditions" do
         subject do
-          create(:course_condition_achievement, :duplicate_child, course: course).tap do |cca|
-            allow(cca).to receive(:achievement_id_changed?).and_return(true)
+          create(:course_condition_achievement, course: course).tap do |achievement_condition|
+            allow(achievement_condition).to receive(:achievement_id_changed?).and_return(true)
           end
         end
         it { is_expected.to_not be_valid }
```
Comment (lowjoel) [inline spec/models/course/condition/achievement_spec.rb]: should check that the field has non-empty errors.

File: spec/models/course/condition/achievement_spec.rb [hunk 4]
```diff
@@ -20,12 +22,26 @@
...
       end
+
+      context 'when an achievement is required by another conditional with the same id' do
+        subject do
+          id = Time.now.to_i
```
Comment (lowjoel) [inline spec/models/course/condition/achievement_spec.rb]: can you create a duplicate without using the time?

Comment (minhtule) [inline spec/models/course/condition/achievement_spec.rb]: Sometimes the id is already created in the database so I need something unique.

File: app/models/course/condition/achievement.rb (hunk 2 above)
Comment (minhtule) [inline app/models/course/condition/achievement.rb]: What is this?

Comment (lowjoel) [inline app/models/course/condition/achievement.rb]: @minhtule do you know what heredoc is?

Comment (minhtule) [inline app/models/course/condition/achievement.rb]: Is it this [one](http://makandracards.com/makandra/1675-using-heredoc-for-prettier-ruby-code)?

Comment (lowjoel) [inline app/models/course/condition/achievement.rb]: Yep. If you use <<-SQL RubyMine will syntax highlight SQL.

File: spec/models/course/condition/achievement_spec.rb (hunk 4 above)
Comment (lowjoel) [inline spec/models/course/condition/achievement_spec.rb]: But time is not unique, either.

Comment (minhtule) [inline spec/models/course/condition/achievement_spec.rb]: Yeah, but timestamp is a big integer so less chance to crash. Any suggestion how to find an available id in test?

**Comment by minhtule** [comment]
@lowjoel please review!

File: spec/models/course/condition/achievement_spec.rb [hunk 5]
```diff
@@ -20,11 +22,29 @@
...
         end
-        it { is_expected.to_not be_valid }
+
+        it 'is not valid' do
+          expect(subject).to_not be_valid
+          expect(subject.errors[:achievement]).to_not be_blank
+        end
+      end
+
+      context 'when an achievement is required by another conditional with the same id' do
```
Comment (lowjoel) [inline spec/models/course/condition/achievement_spec.rb]: @minhtule I stared at this for a while, I finally get what you're trying to do.

Achievement conditions store a reference to their conditional, and the conditional is a polymorphic relation. So you want to make sure that two different polymorphic types with the same ID can exist. Is that what you're testing?

If that's what you're testing, what you should do is make the conditional_type and conditional_id columns a unique index, then this scenario doesn't really need to be tested.

Comment (minhtule) [inline spec/models/course/condition/achievement_spec.rb]: Making (conditional_type, conditional_id) a unique index in the general condition table means the conditional appears only once in the table, which mean a conditional can only have 1 condition?

Comment (lowjoel) [inline spec/models/course/condition/achievement_spec.rb]: Oh, then (conditional_type, conditional_id, condition_type, condition_id) as the unique key? Is that what you're trying to do?

Comment (lowjoel) [inline spec/models/course/condition/achievement_spec.rb]: Wait, hang on, is there a general condition table? I remember only specific conditional condition tables?

Comment (lowjoel) [inline spec/models/course/condition/achievement_spec.rb]: Oh, course_conditions -- it's an actable table, so the unique key should be (actable_id, actable_type, conditional_id, conditional_type). Does that make sense?

Comment (minhtule) [inline spec/models/course/condition/achievement_spec.rb]: Yes, we have a general  condition table and it has 2 polymorphic relations: conditionals and specific conditions. Sure, it's definitely a better solution.

Comment (minhtule) [inline spec/models/course/condition/achievement_spec.rb]: Actually I just recall that this test is to catch the mistake in the SQL query fixed in this PR [here](https://github.com/Coursemology/coursemology2/pull/693/files#diff-bf3adb0382fd1dc7ad1781f2636a1007L40). Please see the comment over there.

Adding the index will le us remove the validation of duplicated specific conditions though.

File: app/models/course/condition/achievement.rb [hunk 6]
```diff
@@ -32,13 +33,16 @@ def achievement_conditions(achievement)
...
     # Workaround, pending the squeel bugfix (activerecord-hackery/squeel#390) that will allow
     # allow the above query to work without #reload
-    Course::Achievement.joins(
-      "INNER JOIN
+    Course::Achievement.joins(<<-SQL)
+      INNER JOIN
         (SELECT cca.achievement_id
           FROM course_condition_achievements cca INNER JOIN course_conditions cc
             ON cc.actable_type = 'Course::Condition::Achievement' AND cc.actable_id = cca.id
-            WHERE cc.conditional_id = #{achievement.id}) ids
```
Comment (minhtule) [inline app/models/course/condition/achievement.rb]: This query gives false validation when 2 different conditionals with the same id have the same condition e.g. when achievement 2 requires achievement 1 and assessment 2 also requires achievement 1. This validation fails (while it's not supposed to). The test catches this error.

File: spec/models/course/condition/achievement_spec.rb (hunk 5 above)
Comment (lowjoel) [inline spec/models/course/condition/achievement_spec.rb]: Okay, I see the problem.

Mark this extra test as a TODO, removing when activerecord-hackery/squeel#390 is fixed, just like the SQL query. Typically you'd be using squeel and not writing raw SQL.

**Comment by minhtule** [comment]
@lowjoel added TODO in the spec

**Pull request merged**